README.md
# **IPK Project 1**

## Description

The project is an implementation of a client for simple calculator server. The client and server can communicate via TCP or UDP. The protocol is described in the assignment repository.

## Usage

The client can be compiled using `make` command. The binary is called `ipkcpc`. The client can be run using `./ipkcpc` command. The client should be run with the following arguments:

- `-h` - hostname of the server (either IPv4 address or hostname)
- `-p` - port of the server
- `-m` - mode of communication, either `tcp` or `udp`

## Implementation

The client is implemented in C++ in the C++20 standard. All of the code is in the `ipkcpc.cpp` source file. Everything is documented to the best detail, so this documentation will only describe the higher level structure and functionality of the code, for a more detailed explanation of parts of the code, please refer to the source code.

### Main function

Main function first parses command line arguments with `parse_args()` function into a global variable of `args_t` type. The arguments are checked if they are valid — non-empty hostname, valid port (range 1024-65535) and valid mode (either `tcp` or `udp`). If any of the arguments is invalid, the program exits with an error message.

Depending on the mode of communication, the appropriate function is called. If the mode is `tcp`, the `tcp_client()` function is called. If the mode is `udp`, the `udp_client()` function is called.

### TCP client

The TCP client first finds the server using the `gethostbyname2()` function. The server is then stored in the `server` structure. The structure is then used to open a TCP socket with the `socket()` function. The socket is then connected to the server with the `connect()` function. This connection lasts for the whole duration of the program. A buffer is created to store both sent and received data. Buffer of size `TCP_BUFSIZE` is created, which is 1024 bytes which should be more than enough for this application.

The program operates in a loop. First it gets the input from the user. This data is sent to the server as text with the `send()` function. The client then waits for a response from the server with the `recv()` function. The response is then printed to the user. The loop starts over.

The client can be terminated by sending `BYE` or pressing `Ctrl+C`, which also sends the `BYE` message to the server.
If the server's response is `BYE`, the client closes the socket, thus terminating the connection, and exits the program.


### UDP client

The UDP client finds the server and creates the structure `server` the same as the TCP client. Then it opens a UDP socket. There are two buffers, one of `UDP_BUFSIZE`, which is 255 bytes, given by the limitation of the protocol which sends the length encoded on 8 bits. The payload is 257 bytes to account for the opcode and data length information.

The client also operates in a loop. First, it gets input from the user. The 0th byte is the operation code, in the request it is set to `0`. The 1st byte is the length of the data following, calculated as `strlen()` of the buffer. The data is sent to the server with the `sendto()` function. The client then waits for a response from the server with the `recvfrom()` function. The 1st byte of the response is the status code, which says whether to print `OK:` or `ERR:`, then the result is printed. The loop starts over.

The client can be terminated by pressing `Ctrl+C`, which closes the socket, thus terminating the connection, and exits the program.

### Error handling

Client prints the appropriate error message standard error output and exits with the code 1. In case of server errors, the client prints the error message from the server, but exits with the code 0.

### Signal handling

The `Ctrl+C` signal is handled by the `sighandler()` function. The code is based on the [answer by Gab Royer on Stack Overflow][3]. The function sets the `continue_loop` global variable to `true`, which is then checked in the main loop. Then, the socket is closed. In TCP mode, the `BYE` message is sent to the server.

## Limitations

The client source code is using libraries only found on Unix systems, therefore, the code cannot be compiled on Windows. On Windows instead of `sys/socket.h`, `arpa/inet.h` and `netinet/in.h`, the `winsock2.h` library should be used which offers similar functionality.

## Testing

The client was tested both on macOS and the provided reference Linux VM against the reference server running on the VM. Testing was made by hand, running all the test cases described in the `testing.md` file, as I was unable to run a Python test script on the VM. Test cases are for basic functionality of error handling, UDP and TCP communication. On top of this, after implementing some change, the client was tested again with many inputs to make sure that the change did not break any functionality, including edge case inputs.

## References

[1]: KOUTENSKÝ, Michal. IPK-Projekty/Project 1 at master - IPK-Projekty - FIT - VUT Brno - git [online]. [cit. 2023-03-21]. Avaiable at: https://git.fit.vutbr.cz/NESFIT/IPK-Projekty/src/branch/master/Project%201

[2]: KOUTENSKÝ, Michal. NESFIT/IPK-Projekty - IPK-Projekty - FIT - VUT Brno - git [online]. [cit. 2023-03-21]. Avaiable at: https://git.fit.vutbr.cz/NESFIT/IPK-Projekty/

[3]: ROYER, Gab. C++ - How can I catch a ctrl-c event? - Stack Overflow [online]. Oct 29, 2009 [cit. 2023-03-21]. Avaiable at: https://stackoverflow.com/a/1641223

[4]: Linux Howtos: C/C++ -> Sockets Tutorial [online]. [cit. 2023-03-21]. Available at: https://www.linuxhowtos.org/C_C++/socket.htm

changelog.md
# Changelog

All notable changes over time will be documented in this file.

Links lead to corresponding commits on Gitea.

I am not aware of any requirements that were not implemented.

---

[2023/03/08](https://git.fit.vutbr.cz/xstrel03/ipk-project-1/commit/c79a08a261e9a746731f32e48feb48c696de2440)
feat: parse args & add Makefile 

[2023/03/15](https://git.fit.vutbr.cz/xstrel03/ipk-project-1/commit/5c06b8f551e71f7c704685bd9a4d1ebeadcd17f1)
feat: add TCP client from stubs

[2023/03/15](https://git.fit.vutbr.cz/xstrel03/ipk-project-1/commit/9542caa247560fd5405aee7a2cf98ec3bd592989)
feat: add UDP client from stubs 

[2023/03/19](https://git.fit.vutbr.cz/xstrel03/ipk-project-1/commit/d687382ab683857f0bfefef6970716b971827757)
feat: cleanup UDP client code, remove broken TCP client

[2023/03/19](https://git.fit.vutbr.cz/xstrel03/ipk-project-1/commit/579505e9daf1a223b6c2171ae481c3cb15aa6343)
feat: add back TCP client code

[2023/03/19](https://git.fit.vutbr.cz/xstrel03/ipk-project-1/commit/60850f94519438f959b5aa4f20843d434fc2e95e)
feat: TCP works, get full message until `\n` 

[2023/03/20](https://git.fit.vutbr.cz/xstrel03/ipk-project-1/commit/17979809560f38a1dd1f667b5abdc8ff5309e4b7)
feat: check for `BYE` and close socket 

[2023/03/20](https://git.fit.vutbr.cz/xstrel03/ipk-project-1/commit/d9b0418074d4346d5e6ab1c26c3e12507304d115)
feat: add signal handling & bigger buffer for TCP 

[2023/03/20](https://git.fit.vutbr.cz/xstrel03/ipk-project-1/commit/b4c179c3bf4f97a713593d7ec3c6bbb7c6f76bf6)
doc: add readme, changelog, tests

ipkcpc.cpp
/**
 * @file ipkcpc.cpp
 *
 * @name IPK project
 * @brief Remote calculator UDP or TCP client
 *
 * @author Matyas Strelec (xstrel03)
 * @date 2023-03-12
 */

 // Cross-platform includes, will work on both Unix and Windows
#include <iostream>
#include <string>
#include <regex>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Unix-specific includes, will work only on Linux, macOS, etc.
// Would need to be changed for Windows (Along with the rest of the code)
#include <signal.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <unistd.h>

#define UDP_BUFSIZE 255
#define TCP_BUFSIZE 1024

// No need to write std:: for everything
using namespace std;

/**
 * @brief Struct for parsed command-line arguments
 *
 */
struct args_t
{
    string host; // IPv4 address of the server
    string port; // Port of the server
    string mode; // 'tcp' or 'udp'
};

// Global variables
args_t args;
bool continue_loop = true;
int client_socket;

/**
 * @brief Close connection and exit
 *
 * If the mode is TCP, send 'BYE' to the server and receive the response
 */
void close_connection() {
    if (args.mode == "tcp") {
        // Send 'BYE' to the server
        send(client_socket, "BYE\n", 4, 0);

        // Receive the response
        char buffer[TCP_BUFSIZE];
        recv(client_socket, buffer, TCP_BUFSIZE, 0);
        cout << buffer;
    }

    close(client_socket);
    exit(EXIT_SUCCESS);
}

/**
 * @brief Signal handler
 *
 * @cite https://stackoverflow.com/a/1641223
 *
 * @param sig Signal number
 */
void sighandler(int sig) {
    continue_loop = false;
    close_connection();
    return;
}

/**
 * @brief Parses command-line arguments
 *
 * Usage: ipkcpc -h <host> -p <port> -m <mode>
 *
 * @param argc Number of arguments
 * @param argv Array of arguments
 * @return Struct containing parsed arguments
 */
args_t parse_args(int argc, char** argv) {
    args_t args;

    for (int i = 1; i < argc; i++) {
        if (string(argv[i]) == "-h") {
            args.host = string(argv[i + 1]);
        }
        else if (string(argv[i]) == "-p") {
            args.port = string(argv[i + 1]);
        }
        else if (string(argv[i]) == "-m") {
            args.mode = string(argv[i + 1]);
        }
    }

    return args;
}

/**
 * @brief TCP client
 *
 * @cite https://git.fit.vutbr.cz/NESFIT/IPK-Projekty/src/branch/master/Stubs/cpp/DemoTcp
 * @cite https://linux.die.net/man/3/gethostbyname2
 *
 * @param host Hostname or address of the server
 * @param port Port number of the server
 * @return Return code
 */
int tcp_client(string host, string port) {
    const char* server_hostname = host.c_str(); // Hostname of the server
    int port_number = atoi(port.c_str());       // Port number of the server
    int bytestx;                                // Bytes sent
    int bytesrx;                                // Bytes received
    struct hostent* server;                     // Server structure
    struct sockaddr_in server_address;          // Server address

    // Get the server IPv4 address
    server = gethostbyname2(server_hostname, AF_INET);
    if (server == NULL) {
        cout << "Error: No such host as " << server_hostname << endl;
        return EXIT_FAILURE;
    }

    // Set the server address
    // Zero the structure
    bzero((char*)&server_address, sizeof(server_address));
    // Set the address family, AF_INET for IPv4
    server_address.sin_family = AF_INET;
    // Copy the server address to the structure
    bcopy((char*)server->h_addr, (char*)&server_address.sin_addr.s_addr, server->h_length);
    // Set the port number
    server_address.sin_port = htons(port_number);

    // Create the socket
    client_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (client_socket <= 0) {
        perror("Error: Socket");
        return EXIT_FAILURE;
    }

    // Connect to the server
    int connect_result = connect(client_socket, (struct sockaddr*)&server_address, sizeof(server_address));
    if (connect_result < 0) {
        perror("Error: Connect");
        return EXIT_FAILURE;
    }

    char buffer[TCP_BUFSIZE];       // Buffer for the calculation data

    while (continue_loop) {
        // Clear the buffer
        bzero(buffer, TCP_BUFSIZE);

        // Get the data from the user
        fgets(buffer, TCP_BUFSIZE - 1, stdin);

        // Send the data to the server
        bytestx = send(client_socket, buffer, strlen(buffer), 0);
        if (bytestx < 0) {
            perror("Error: Send");
            close(client_socket);
            return EXIT_FAILURE;
        }
        // Clear the buffer
        bzero(buffer, TCP_BUFSIZE);

        // Receive the data from the server
        bytesrx = recv(client_socket, buffer, TCP_BUFSIZE, 0);
        if (bytesrx < 0) {
            perror("Error: Receive");
            close(client_socket);
            return EXIT_FAILURE;
        }

        // Print the result
        cout << buffer;

        // Check for 'BYE' message
        if (buffer[0] == 'B' && buffer[1] == 'Y' && buffer[2] == 'E') {
            continue_loop = false;
        }
    }

    close(client_socket);

    return 0;
}

/**
 * @brief UDP client
 *
 * @cite https://git.fit.vutbr.cz/NESFIT/IPK-Projekty/src/branch/master/Stubs/cpp/DemoUdp
 * @cite https://linux.die.net/man/3/gethostbyname2
 *
 * @param host Hostname or address of the server
 * @param port Port number of the server
 * @return Return code
 */
int udp_client(string host, string port) {
    const char* server_hostname = host.c_str(); // Hostname of the server
    int port_number = atoi(port.c_str());       // Port number of the server
    int bytestx;                                // Bytes sent
    int bytesrx;                                // Bytes received
    socklen_t server_len;                       // Length of the server address
    struct hostent* server;                     // Server structure
    struct sockaddr_in server_address;          // Server address

    // Get the server IPv4 address
    server = gethostbyname2(server_hostname, AF_INET);
    if (server == NULL) {
        cout << "Error: No such host as " << server_hostname << endl;
        return EXIT_FAILURE;
    }

    // Set the server address
    // Zero the structure
    bzero((char*)&server_address, sizeof(server_address));
    // Set the address family, AF_INET for IPv4
    server_address.sin_family = AF_INET;
    // Copy the server address to the structure
    bcopy((char*)server->h_addr, (char*)&server_address.sin_addr.s_addr, server->h_length);
    // Set the port number
    server_address.sin_port = htons(port_number);

    // Create the socket
    client_socket = socket(AF_INET, SOCK_DGRAM, 0);
    if (client_socket <= 0) {
        perror("Error: Socket");
        return EXIT_FAILURE;
    }

    char buffer[UDP_BUFSIZE];       // Buffer for the calculation data
    char payload[UDP_BUFSIZE + 2];  // Buffer for the payload (opcode + data length + calculation data)


    while (continue_loop) {
        // Clear the buffers
        bzero(buffer, UDP_BUFSIZE);
        bzero(payload, UDP_BUFSIZE + 2);

        // Get the payload data from the user
        cin.getline(buffer, UDP_BUFSIZE);

        // Set the opcode to 0 for request
        payload[0] = 0;

        // Get the length of the payload data
        // Set the payload length in the buffer
        int buffer_length = strlen(buffer);
        payload[1] = (char)buffer_length;

        // Copy the buffer data to the payload
        strcpy(&payload[2], buffer);

        // Send the data to the server
        server_len = sizeof(server_address);
        bytestx = sendto(client_socket, payload, buffer_length + 2, 0, (struct sockaddr*)&server_address, server_len);
        if (bytestx < 0) {
            perror("ERROR: Sendto");
            close(client_socket);
            return EXIT_FAILURE;
        }

        // Clear the buffers
        bzero(buffer, UDP_BUFSIZE);
        bzero(payload, UDP_BUFSIZE + 2);

        // Receive the data from the server
        bytesrx = recvfrom(client_socket, buffer, UDP_BUFSIZE, 0, (struct sockaddr*)&server_address, &server_len);
        if (bytesrx < 0) {
            perror("ERROR: Recvfrom");
            close(client_socket);
            return EXIT_FAILURE;
        }

        // Get the status of the response
        char out_status = buffer[1];

        if (out_status == 0)
            cout << "OK:" << buffer + 3 << endl;
        else if (out_status == 1)
            cout << "ERR:" << buffer + 3 << endl;
    }

    close(client_socket);
    return 0;
}


/**
 * @brief Main function
 *
 * @param argc Number of arguments
 * @param argv Array of arguments
 * @return Exit code
 */
int main(int argc, char** argv) {
    // Load arguments to global variable
    // To be able to close conenction correctly on signal
    args = parse_args(argc, argv);

    // Check if host is correct
    if (args.host.empty()) {
        cerr << "Error: Host is not specified." << endl;
        return 1;
    }

    // Check if port is in range
    if (!(stoi(args.port) > 1024 && stoi(args.port) < 65535)) {
        cerr << "Error: Port must be in range 1024-65535." << endl;
        return 1;
    }

    signal(SIGINT, &sighandler);

    // Select mode based on the parsed arguments
    // Handle error if mode is not 'tcp' or 'udp'
    if (args.mode == "tcp")
        tcp_client(args.host, args.port);
    else if (args.mode == "udp")
        udp_client(args.host, args.port);
    else {
        cerr << "Error: Unknown mode '" << args.mode << "', must be either 'udp' or 'tcp'." << endl;
        return 1;
    }
}